### ISSUE TITLE
Possible infinite loop in findPrevious()/findNext()

### ISSUE THREAD

**Issue by niconoe-** [squizlabs/PHP_CodeSniffer#1206]
Hi developers,

I'm using PHPCS since a while and I just encountered an issue with an infinite loop by using the method `\PHP_CodeSniffer_File::findPrevious`. 

This method loops over tokens positions from `$start` and rewind to `$end` in a `for` loop. But I experienced, (probably writing my own sniff) that when setting `$start` to a negative number and `$end` to a boolean, you trigger an infinite loop.

`for ($i = $start; $i >= $end; $i--);`

Here, a numerical comparison is required between `$i` and `$end`, but PHP does something that might be seen as unexpected behavior at first sight. As `$end` is a boolean, it transtypes `$i` as a boolean first for comparison. Then, the comparision is made with **integer values of those booleans**.

This means that if `$i = -8` and `$end = false`, the really comparison will be [-8 -> true -> 1] >= [false -> 0], which is actually `true` while you're expecting that `-8 >= 0` is `false`.

As you understand now easily, decrementing `$i` will change nothing about its true value when comparating, and so we're creating an infinite loop.

To fix this issue, you must transtype `$start` and `$end` to integers, so the comparison will not implicitly transtype `$i` in boolean just before comparing. The implicit transtyping will be done in integer, which is the type we want.

This problem also occurs on `\PHP_CodeSniffer_File::findNext`, as the loop is reversed but the same.

Thank you for making PHP_CodeSniffer better every day!
Best regards,

**Comment by gsherwood** [comment]
I'm sure you are right, but you are completely misusing the method. Adding additional checks and processing time for such heavily used methods isn't something I want to do, just to protect against misuse of the method by developers.

`$start` is the position in the token stack you want to start looking from. The token stack starts at 0, so PHPCS will never give you a negative number to use. Only manually passing in a completely invalid value will cause it to be negative.

`$end` is the position in the token stack you want to stop looking. Again, it should be a non-negative integer and PHPCS will never give you an invalid value to pass into this method. If you don't want to set an end point, you pass `null` - not `false` - just like you do for `findNext()`.

If you have a legitimate use case where this came up, please describe it for me so I can see if PHPCS can support it. If you are just trying to break the methods, thank you for your feedback, but I suggest you use them as intended so you don't get an infinite loop.

**Comment by niconoe-** [comment]
Thank you for your feedback.

You are right, the method is totally misused in my own sniff, that's why I produce the infinite loop. 
But I keep thinking this could be a trouble for developers that are writing their own sniffs.

My case was just the seek of a specific token, but in a loop as I wanted to step back until the one I wanted (I mean with specific information). 

The trouble I mainly have when writting my own sniffs is when I need to findPrevious or findNext, if I want the same category of token I'm looking for, but not the one at the position I found so far, I need to call `findPrevious($token, $ptr - 1)` or `findNext($token, $ptr + 1)`, otherwise, the `find*` method will return exactly the same token (`$token`).

In my sniff, I encountered a case where, due to other findPrevious/findNext, my `$ptr` was `false`, resulting in some faulty call, like  `findPrevious($token, -1)` which is actually wrong. I just test that my `$ptr` is not false and managed this case, and avoid the infinite loop.

Sorry for that, I understand that, for a tool like PHP_CodeSniffer, the performance cannot be reduce to solve problems like this infinite loop ;).

Have a nice day.
Regards,
